custom_components/tesla_custom/sensor.py

```diff
@@ -1,112 +1,340 @@
+        self.attrs.update(data)
+        return self.attrs
+
+
+class TeslaChargerEnergy(TeslaCarDevice, SensorEntity):
+    """Representation of the Tesla Energy Added."""
+
+    def __init__(
+        self, hass: HomeAssistant, car: dict, coordinator: TeslaDataUpdateCoordinator
+    ) -> None:
+        """Initialize the Sensor Entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "energy added"
+        self._attr_device_class = SensorDeviceClass.ENERGY
+        self._attr_state_class = SensorStateClass.TOTAL_INCREASING
+        self._attr_native_unit_of_measurement = ENERGY_KILO_WATT_HOUR
+        self._attr_icon = "mdi:lightning-bolt"
+
+    @property
+    def native_value(self) -> int:
+        """Return the Charge Energy Added."""
+        charge_energy_added = self.car.charging.get("charge_energy_added")
+
+        return charge_energy_added
+
+
+class TeslaMileage(TeslaCarDevice, SensorEntity):
+    """Representation of the Tesla Energy Added."""
+
+    def __init__(
+        self, hass: HomeAssistant, car: dict, coordinator: TeslaDataUpdateCoordinator
+    ) -> None:
+        """Initialize the Sensor Entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "mileage"
+        self._attr_device_class = None
+        self._attr_state_class = SensorStateClass.TOTAL_INCREASING
+        self._attr_icon = "mdi:counter"
+
+    @property
+    def native_value(self) -> int:
+        """Return the Charge Energy Added."""
+        odometer_value = self.car.state.get("odometer")
+
+        if odometer_value is None:
+            return None
+
+        if self.native_unit_of_measurement == LENGTH_KILOMETERS:
+            odometer_value = round(
+                convert(odometer_value, LENGTH_MILES, LENGTH_KILOMETERS), 2
             )
-        return attr
+
+        return round(odometer_value, 2)
+
+    @property
+    def native_unit_of_measurement(self) -> str:
+        """Return Native Unit of Measurement.
+
+        Get from Car setting
+        """
+        gui_uom = self.car.gui.get("gui_distance_units", "mi/hr")
+
+        if gui_uom == "mi/hr":
+            return LENGTH_MILES
+
+        return LENGTH_KILOMETERS
+
+
+class TeslaRange(TeslaCarDevice, SensorEntity):
+    """Representation of the Tesla Energy Added."""
+
+    def __init__(
+        self, hass: HomeAssistant, car: dict, coordinator: TeslaDataUpdateCoordinator
+    ) -> None:
+        """Initialize the Sensor Entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "range"
+        self._attr_device_class = None
+        self._attr_state_class = SensorStateClass.MEASUREMENT
+        self._attr_icon = "mdi:gauge"
+
+    @property
+    def native_value(self) -> int:
+        """Return the Charge Energy Added."""
+        range_value = self.car.charging.get("battery_range")
+
+        if self.car.gui.get("gui_range_display") == "Rated":
+            range_value = self.car.charging.get("ideal_battery_range")
```

Comment:
This is incorrect for my 2015 Model S 85D. If `gui_range_display` is set to `Rated`, it should be returning `battery_value`, not `ideal_battery_range`. That should be returned if `gui_range_display` is set to `Ideal`.

Might be some differences depending on model, year, country, etc. so I haven't changed it yet. Would like to get some more feedback first.

Reply:
I set this to return rated range when the GUI option is set to rated range. I can't think of any reason as to why ideal range would be returned in this case.